### PR TITLE
Fix DecayFunctions' toString

### LINKED ISSUE #100870: Exponential, Gauss and Linear decay function classes have incorrect `equals`
### Elasticsearch Version

8.10.3

### Installed Plugins

_No response_

### Java Version

Java 17

### OS Version

Windows 10

### Problem Description

The `equals` method implementation of the following classes looks incorrect:
- `ExponentialDecayFunctionBuilder.ExponentialDecayScoreFunction`
- `GaussDecayFunctionBuilder.GaussScoreFunction`
- `LinearDecayFunctionBuilder.LinearDecayScoreFunction`

They all look like this:
```java
public boolean equals(Object obj) {
    if (super.equals(obj)) {
        return true;
    }
    return obj != null && getClass() != obj.getClass();
}
```

There are two problems:
- The fast `return true` for `super.equals(obj)` seems quite brittle and only works because those classes all directly extend `Object` and therefore that `super.equals` call actually checks for identity equality
- `getClass() != obj.getClass()` looks wrong; that should probably check `==`

Suggested implementation:
```java
public boolean equals(Object obj) {
    return obj == this
        || (obj != null && obj.getClass() == this.getClass());
}
```

### Steps to Reproduce

These all erroneously print "true":
```java
System.out.println(ExponentialDecayFunctionBuilder.EXP_DECAY_FUNCTION.equals("a string"));
System.out.println(GaussDecayFunctionBuilder.GAUSS_DECAY_FUNCTION.equals("a string"));
System.out.println(LinearDecayFunctionBuilder.LINEAR_DECAY_FUNCTION.equals("a string"));
```

### Logs (if relevant)

_No response_

### PR DESCRIPTION
**Pull request by tteofili**
Trivial fix for `DecayFunction#toString` implementors.
 
closes #100870

### CODE REVIEW

File: server/src/internalClusterTest/java/org/elasticsearch/search/functionscore/FunctionScorePluginIT.java [hunk 1]
```diff
@@ -146,7 +146,6 @@ private static class LinearMultScoreFunction implements DecayFunction {
 
             @Override
             public double evaluate(double value, double scale) {
-
```
Comment (carlosdelest) [inline server/src/internalClusterTest/java/org/elasticsearch/search/functionscore/FunctionScorePluginIT.java:149]: Spacing nit: We probably don't need this change.

**Review by carlosdelest — APPROVED** [review]
💯

File: server/src/internalClusterTest/java/org/elasticsearch/search/functionscore/FunctionScorePluginIT.java (hunk 1 above)
Comment (tteofili) [reply server/src/internalClusterTest/java/org/elasticsearch/search/functionscore/FunctionScorePluginIT.java:149]: I know, it is just aesthetics :)

**Comment by benwtrent** [comment]
@tteofili do we think this is ready to merge?

File: docs/changelog/107415.yaml [hunk 2]
```diff
@@ -0,0 +1,6 @@
+pr: 107415
+summary: Fix `DecayFunctions'` `toString`
```
Comment (Marcono1234) [inline docs/changelog/107415.yaml:2]: Why does this say `toString`? Unless I am overlooking something this only affects `equals`, and `toString` is not affected at all.

(Thanks though for fixing the issue with the `equals` implementations!)

**Pull request merged**